src/main/java/gov/cms/mat/cql_elm_translation/service/CqlConversionService.java

```diff
@@ -180,4 +198,138 @@ private String formatMessage(CqlCompilerException e) {
+    if (usings != null) {
+      List<UsingDef> defs = usings.getDef();
+      for (UsingDef def : defs) {
+        List<TrackBack> trackBacks = def.getTrackbacks();
+        if (CollectionUtils.isNotEmpty(trackBacks)) {
+          for (TrackBack trackBack : trackBacks) {
+            endLines.add(trackBack.getEndLine());
+          }
+        }
+      }
+    }
+    return endLines;
+  }
+
+  protected List<Integer> getParameterEndLines(CqlTranslator cqlTranslator) {
+    List<Integer> endLines = new ArrayList<>();
+    Library.Parameters parameters =
+        cqlTranslator.getTranslatedLibrary().getLibrary().getParameters();
+    if (parameters != null) {
+      List<ParameterDef> defs = parameters.getDef();
+      for (ParameterDef def : defs) {
+        List<TrackBack> trackBacks = def.getTrackbacks();
+        if (CollectionUtils.isNotEmpty(trackBacks)) {
+          for (TrackBack trackBack : trackBacks) {
+            endLines.add(trackBack.getEndLine());
+          }
+        }
+      }
+    }
+    return endLines;
+  }
+
+  protected List<Integer> getValueSetsEndLines(CqlTranslator cqlTranslator) {
+    List<Integer> endLines = new ArrayList<>();
+    Library.ValueSets valuesets = cqlTranslator.getTranslatedLibrary().getLibrary().getValueSets();
+    if (valuesets != null) {
+      List<ValueSetDef> defs = valuesets.getDef();
+      for (ValueSetDef def : defs) {
+        List<TrackBack> trackBacks = def.getTrackbacks();
+        if (CollectionUtils.isNotEmpty(trackBacks)) {
+          for (TrackBack trackBack : trackBacks) {
+            endLines.add(trackBack.getEndLine());
+          }
+        }
+      }
+    }
+    return endLines;
+  }
+
+  protected List<Integer> getCodeSystemEndLines(CqlTranslator cqlTranslator) {
+    List<Integer> endLines = new ArrayList<>();
+    Library.CodeSystems codeSystems =
+        cqlTranslator.getTranslatedLibrary().getLibrary().getCodeSystems();
+    if (codeSystems != null) {
+      List<CodeSystemDef> defs = codeSystems.getDef();
+      for (CodeSystemDef def : defs) {
+        List<TrackBack> trackBacks = def.getTrackbacks();
+        if (CollectionUtils.isNotEmpty(trackBacks)) {
+          for (TrackBack trackBack : trackBacks) {
+            endLines.add(trackBack.getEndLine());
+          }
+        }
+      }
+    }
+    return endLines;
+  }
+
+  protected List<Integer> getCodeEndLines(CqlTranslator cqlTranslator) {
+    List<Integer> endLines = new ArrayList<>();
+    Library.Codes codes = cqlTranslator.getTranslatedLibrary().getLibrary().getCodes();
+    if (codes != null) {
+      List<CodeDef> defs = codes.getDef();
+      for (CodeDef def : defs) {
+        List<TrackBack> trackBacks = def.getTrackbacks();
+        if (CollectionUtils.isNotEmpty(trackBacks)) {
+          for (TrackBack trackBack : trackBacks) {
+            endLines.add(trackBack.getEndLine());
+          }
+        }
+      }
+    }
+    return endLines;
+  }
```

Comment:
the closer look at this gives impression that this can be simplified with just few lines and reused for all line methods

```
protected List<Integer> getLines(List<? extends Element> defs) {
    List<Integer> startLines = new ArrayList<>();
    if (CollectionUtils.isNotEmpty(defs)) {
      for (Element def : defs) {
        List<TrackBack> trackBacks = def.getTrackbacks();
        if (CollectionUtils.isNotEmpty(trackBacks)) {
          for (TrackBack trackBack : trackBacks) {
            startLines.add(trackBack.getStartLine());
          }
        }
      }
    }
    return startLines;
  }
 
```

Reply:
@adongare done, thanks~